community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/commands/expressions/DistanceFunction.scala

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (c) 2002-2015 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher.internal.compiler.v3_0.commands.expressions
+
+import org.neo4j.cypher.internal.compiler.v3_0.ExecutionContext
+import org.neo4j.cypher.internal.compiler.v3_0.pipes.QueryState
+import org.neo4j.cypher.internal.compiler.v3_0.symbols.SymbolTable
+import org.neo4j.cypher.internal.frontend.v3_0.symbols._
+import Math._
+
+case class DistanceFunction(p1: Expression, p2: Expression) extends Expression {
+
+  def apply(ctx: ExecutionContext)(implicit state: QueryState): Any = {
+    val geometry1 = convertToGeometry(p1, ctx);
+    val geometry2 = convertToGeometry(p2, ctx);
+    // TODO: Support better calculations, like https://en.wikipedia.org/wiki/Vincenty%27s_formulae
+    // TODO: Support more coordinate systems
+    HaversinCalculator.distance(geometry1, geometry2).getOrElse(
+      CartesianCalculator.distance(geometry1, geometry2).getOrElse(
+        throw new IllegalArgumentException(s"Invalid points passed to distance($p1, $p2)")
+      )
+    )
+  }
+
+  def convertToGeometry(p: Expression, ctx: ExecutionContext)
+                       (implicit state: QueryState): Geometry = p match {
+    case Identifier(name) => ctx(name) match {
+      case p: PointFunction => p(ctx).asInstanceOf[Geometry]
+      case m: Geometry => m
+    }
+    case p: PointFunction => p(ctx).asInstanceOf[Geometry]
+  }
+
+  def rewrite(f: (Expression) => Expression) = f(DistanceFunction(p1.rewrite(f), p2.rewrite(f)))
+
+  def calculateType(symbols: SymbolTable): CypherType = CTFloat
+
+  override def arguments: Seq[Expression] = p1.arguments ++ p2.arguments
+
+  override def symbolTableDependencies: Set[String] = p1.symbolTableDependencies ++ p2.symbolTableDependencies
+
+  override def toString = "Distance(" + p1 + ", " + p2 + ")"
+}
+
+trait DistanceCalculator {
+  def appliesTo(p1: Geometry, p2: Geometry): Boolean
```

Comment:
I suggest a different name than `appliesTo`, simply because `apply` is a different concept in Scala. That said, it might be good to rework this such that it is defined as a PartialFunction, since that's in line with what it's doing (a function that is only defined for certain inputs).